substrate/frame/opf/src/functions.rs

```diff
@@ -0,0 +1,360 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! # Helper functions for OPF pallet.
+
+pub use super::*;
+impl<T: Config> Pallet<T> {
+	pub fn get_formatted_call(call: Call<T>) -> <T as Config>::RuntimeCall {
+		call.into()
+	}
+	pub fn conviction_amount(
+		amount: BalanceOf<T>,
+		conviction: Democracy::Conviction,
+	) -> Option<BalanceOf<T>> {
+		let conviction_amount: BalanceOf<T> = match conviction {
+			Democracy::Conviction::None =>
+				amount.checked_div(&10u8.into()).unwrap_or_else(Zero::zero),
+			_ =>
+				amount.saturating_mul(<u8 as From<Democracy::Conviction>>::from(conviction).into()),
+		};
+		Some(conviction_amount)
+	}
+
+	pub fn create_proposal(
+		caller: T::AccountId,
+		proposal_call: pallet::Call<T>,
+	) -> BoundedCallOf<T> {
+		let proposal = Box::new(Self::get_formatted_call(proposal_call.into()));
+		let value = <T as Democracy::Config>::MinimumDeposit::get();
+		let call = Call::<T>::execute_call_dispatch { caller: caller.clone(), proposal };
+		let call_formatted = Self::get_formatted_call(call.into());
+		let bounded_proposal = <T as Democracy::Config>::Preimages::bound(call_formatted.into())
+			.expect("Operation failed");
+		let _hash = Democracy::Pallet::<T>::propose(
+			RawOrigin::Signed(caller).into(),
+			bounded_proposal.clone(),
+			value,
+		);
+		bounded_proposal
+	}
+
+	pub fn start_dem_referendum(
+		proposal: BoundedCallOf<T>,
+		delay: BlockNumberFor<T>,
+	) -> Democracy::ReferendumIndex {
+		let threshold = Democracy::VoteThreshold::SimpleMajority;
+		let referendum_index =
+			Democracy::Pallet::<T>::internal_start_referendum(proposal, threshold, delay);
+		referendum_index
+	}
+
+	// Helper function for voting action. Existing votes are over-written, and Hold is adjusted
+	pub fn try_vote(
```

Comment:
The function here doesn't check that the voter is not voting with their funds for multiple proposals. If I would have 1 million tokens, I could vote with one million tokens for every proposal. That should not be allowed.

We should add a method `voting(user, vote, index) -> DispatchResult` to the `Polling` trait. The default implementation can just return `Ok(())` and that's it. In our case we would use this to ensure that the funds the user wants to vote with are not already being used for a different project.

This `voting` function should be called [here](https://github.com/paritytech/polkadot-sdk/blob/20ffada06152b8926cbccfa19334fe7bbb116857/substrate/frame/conviction-voting/src/lib.rs#L404) right before we fetch the poll.

Reply:
I don't know  if I properly understand, but the amount commited is locked in `try_vote`, and in the extrinsic `vote`, I always check if the voter unholded balance is enough for him to commit the amount specified : If you have 1 million tokens, It is normally not possible to commit 1 million tokens for every projects.
I hope this makes sense